OrbitVulkanLayer/VulkanLayerProducerImplTest.cpp

```diff
@@ -0,0 +1,512 @@
+#include "VulkanLayerProducerImpl.h"
+
+namespace orbit_vulkan_layer {
+
+namespace {
+
+class MockCaptureStatusListener : public VulkanLayerProducer::CaptureStatusListener {
+ public:
+  MOCK_METHOD(void, OnCaptureStart, (), (override));
+  MOCK_METHOD(void, OnCaptureStop, (), (override));
+  MOCK_METHOD(void, OnCaptureFinished, (), (override));
+};
+
+class VulkanLayerProducerImplTest : public ::testing::Test {
+ protected:
+  void SetUp() override {
+    fake_service_.emplace();
+
+    grpc::ServerBuilder builder;
+    builder.RegisterService(&fake_service_.value());
+    fake_server_ = builder.BuildAndStart();
+    ASSERT_NE(fake_server_, nullptr);
+
+    std::shared_ptr<grpc::Channel> channel =
+        fake_server_->InProcessChannel(grpc::ChannelArguments{});
+
+    producer_.emplace();
+    producer_->SetCaptureStatusListener(&mock_listener_);
+    producer_->BringUp(channel);
+
+    // Leave some time for the ReceiveCommandsAndSendEvents RPC to actually happen.
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+  }
+
+  void TearDown() override {
+    // Leave some time for all pending communication to finish.
+    std::this_thread::sleep_for(std::chrono::milliseconds(50));
+    producer_->TakeDown();
+    producer_->SetCaptureStatusListener(nullptr);
+    producer_.reset();
+
+    fake_service_->FinishAndDisallowRpc();
+    fake_server_->Shutdown();
+    fake_server_->Wait();
+
+    fake_service_.reset();
+    fake_server_.reset();
+  }
+
+  std::optional<orbit_producer::FakeProducerSideService> fake_service_;
+  std::unique_ptr<grpc::Server> fake_server_;
+  std::optional<VulkanLayerProducerImpl> producer_;
+  MockCaptureStatusListener mock_listener_;
+
+  static const std::string kInternedString1;
+  static const uint64_t kExpectedInternedString1Key;
+  static const std::string kInternedString2;
+  static const uint64_t kExpectedInternedString2Key;
+  static const std::string kInternedString3;
+  static const uint64_t kExpectedInternedString3Key;
+};
+
+const std::string VulkanLayerProducerImplTest::kInternedString1 = "a";
+const uint64_t VulkanLayerProducerImplTest::kExpectedInternedString1Key =
+    std::hash<std::string>{}(kInternedString1);
+
+const std::string VulkanLayerProducerImplTest::kInternedString2 = "b";
+const uint64_t VulkanLayerProducerImplTest::kExpectedInternedString2Key =
+    std::hash<std::string>{}(kInternedString2);
+
+const std::string VulkanLayerProducerImplTest::kInternedString3 = "c";
+const uint64_t VulkanLayerProducerImplTest::kExpectedInternedString3Key =
+    std::hash<std::string>{}(kInternedString3);
+
+constexpr std::chrono::duration kWaitMessagesSentDuration = std::chrono::milliseconds(25);
+
+}  // namespace
+
+TEST_F(VulkanLayerProducerImplTest, IsCapturingAndListener) {
+  EXPECT_FALSE(producer_->IsCapturing());
+
+  EXPECT_CALL(mock_listener_, OnCaptureStart).Times(1);
+  fake_service_->SendStartCaptureCommand();
+  std::this_thread::sleep_for(kWaitMessagesSentDuration);
```

Comment:
Could you work around those sleeps, e.g. by setting a flag in the OnCaptureStart/Stop methods?

Reply:
I know it's far from ideal, but using `OnCaptureStart` as an indication to continue kind of interferes with the tests (as that's what we are testing), and if the expectation is not satisfied it means that the test will hang instead of failing. By the way, I'm doing the same in the tests for `ProducerSideService` and `CaptureEventProducer`.

Reply:
Yeah you are right, this is also not ideal. I'd still slightly feel it would be better. Two thoughts on this:
1. By keeping the EXPECT_CALL regarding the OnCaptureStart, you would keep test testing this behaviour. You may need to insert a "validate and reset". 
2. You could still add a timeout, to prevent from a hold. That timeout may be much larger, such that it is at least less likely that it failed falsely.

Reply:
But also note, I am not too strong about this.

Reply:
I see your point. If we need to tackle this, we need to do it together with the tests for `ProducerSideService`, `CaptureEventProducer` and `LockFreeBufferCaptureEventProducer` and give it some good thought. That's because there are quite big drawbacks:

- Every one-line sleep is replaced with much more complicated code (action on listener method, wait-notify, timeout), which complicates the tests even more and makes them even harder to read, as they are already hundreds of lines of code.
- We would need to be sure that every single sleep could be remove. Otherwise the tests would become inconsistent, as there would be two ways to implement a wait. That would mean that the tests would be harder to read, without actually solving the problem fully.

Reply:
Agree. I am fine with tackling this as a follow up. I still think that we can set something up by using the Test class and the MockListener, so that the code complexity will not increase (that much) in the actual tests.
I've done something similar (but in a different context), with the IsCapturing in the SubmissionTrackerTest, before we had the Listener.

Reply:
Let me at least create a bug at least.

Reply:
Thanks

Reply:
http://b/175669610